### PR TITLE
feat: send message editions and handle error state [AR-2900]

### PR DESCRIPTION
**Pull request by saleniuk**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/AR-2900" title="AR-2900" target="_blank">SubtaskAR-2900</a>  Create a new use case to handle message editing
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
    - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
    - [x] contains a reference JIRA issue number like `SQPIT-764`
    - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
    - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

Currently, there is no logic implemented for sending message editions.

### Solutions

Use created use case for sending message editions, change the composer state to inactive after sending the edit (like on designs), add new `MessageStatus.EditSendFailure` which combines both text types (error label and status box).

### Dependencies (Optional)

Needs releases with:

- https://github.com/wireapp/kalium/pull/1583

#### How to Test

Send a message, long-click on that message and choose "edit", modify the content and send it.

### Attachments (Optional)

https://user-images.githubusercontent.com/30429749/225684549-5504631c-3bcc-4310-b5cc-29e0733c31aa.mp4

----
#### PR Post Submission Checklist for internal contributors (Optional)

- [x] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

- [x] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

**Comment by AndroidBob** [comment]
Build [1](https://10.10.124.134/job/wire-android-reloaded/job/PR-1557/1/) **failed**.

**Comment by AndroidBob** [comment]
Build [2](https://10.10.124.134/job/wire-android-reloaded/job/PR-1557/2/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv4.0.2(6001873)(6001873)-staging-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/staging/compat/com.wire.android-vv4.0.2(6001873)(6001873)-staging-compat.apk)

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/model/UIMessage.kt [hunk 1]
```diff
@@ -79,19 +77,43 @@ data class MessageFooter(
...
 )
 
-sealed class MessageStatus(val text: UIText) {
-    object Untouched : MessageStatus(UIText.DynamicString(""))
-    object Deleted : MessageStatus(UIText.StringResource(R.string.deleted_message_text))
-    data class Edited(val formattedEditTimeStamp: String) :
-        MessageStatus(UIText.StringResource(R.string.label_message_status_edited_with_date, formattedEditTimeStamp))
+sealed class MessageStatus(
+    open val errorText: UIText? = null, // error description text shown below the content of the message
+    open val badgeText: UIText? = null, // text shown between tne user name and the content in the outlined box with a text inside
```
Comment (gongracr) [inline app/src/main/kotlin/com/wire/android/ui/home/conversations/model/UIMessage.kt]:
```suggestion
    open val badgeText: UIText? = null, // text shown between the user name and the content in the outlined box with a text inside
```

File: app/src/main/kotlin/com/wire/android/ui/home/messagecomposer/MessageComposer.kt [hunk 2]
```diff
@@ -100,6 +102,24 @@ fun MessageComposer(
...
 
+        val onSendEditButtonClicked = remember {
+            {
+                (messageComposerState.messageComposeInputState as? MessageComposeInputState.Active)?.let {
+                    (it.type as? MessageComposeInputType.EditMessage)?.messageId
+                }?.let { originalMessageId ->
+                    onSendEditTextMessage(
+                        EditMessageBundle(
+                            originalMessageId = originalMessageId,
+                            newContent = messageComposerState.messageComposeInputState.messageText.text,
```
Comment (gongracr) [inline app/src/main/kotlin/com/wire/android/ui/home/messagecomposer/MessageComposer.kt:113]: Should we check that the new edit message bundle text is not empty somehow?

File: app/src/main/kotlin/com/wire/android/ui/home/messagecomposer/MessageComposerInnerState.kt [hunk 3]
```diff
@@ -173,6 +173,15 @@ data class MessageComposerInnerState(
...
         )
+        quotedMessageData = null
+        inputFocusRequester.requestFocus()
+    }
+
+    fun toNewMessage(clearInput: Boolean = false) {
+        messageComposeInputState = MessageComposeInputState.Active(
+            messageText = if (clearInput) TextFieldValue("") else messageComposeInputState.messageText,
+            type = MessageComposeInputType.NewMessage()
+        )
```
Comment (gongracr) [inline app/src/main/kotlin/com/wire/android/ui/home/messagecomposer/MessageComposerInnerState.kt]: Where is this been used?

File: app/src/main/kotlin/com/wire/android/ui/home/messagecomposer/MessageComposer.kt (hunk 2 above)
Comment (gongracr) [reply app/src/main/kotlin/com/wire/android/ui/home/messagecomposer/MessageComposer.kt:113]: But I guess we would need a more finely grained handled case for that from product/design side. Like, should we trigger message deletion when attempting to "empty" a message? should we show a dialog?

**Review by gongracr — APPROVED** [review]
Looking grrrr8 🐯 Just some small comments about empty edited message case. But nothing blocking so approving in advance as I assume it can also be handled later on as well.

File: app/src/main/kotlin/com/wire/android/ui/home/messagecomposer/MessageComposer.kt (hunk 2 above)
Comment (saleniuk) [reply app/src/main/kotlin/com/wire/android/ui/home/messagecomposer/MessageComposer.kt:113]: There's a check already and the "save" button is disabled if you try to enter empty message text

File: app/src/main/kotlin/com/wire/android/ui/home/messagecomposer/MessageComposerInnerState.kt (hunk 3 above)
Comment (saleniuk) [reply app/src/main/kotlin/com/wire/android/ui/home/messagecomposer/MessageComposerInnerState.kt]: Oh right, now it's not, I wanted to use that after the edit is sent, but decided to change to "inactive" state in that case, I will remove this method

**Comment by AndroidBob** [comment]
Build [3](https://10.10.124.134/job/wire-android-reloaded/job/PR-1557/3/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv4.0.2(6007741)(6007741)-staging-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/staging/compat/com.wire.android-vv4.0.2(6007741)(6007741)-staging-compat.apk)

**Review by vitorhugods — APPROVED** [review]
Looks good to me.

Would be nice to cover `MessageComposerViewModel$sendEditMessage` with a couple tests.

**Comment by AndroidBob** [comment]
Build [4](https://10.10.124.134/job/wire-android-reloaded/job/PR-1557/4/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv4.0.2(6010034)(6010034)-staging-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/staging/compat/com.wire.android-vv4.0.2(6010034)(6010034)-staging-compat.apk)

**Pull request merged**